shortest_common_superstring/paluch_elbassoni_zuylen.py

```diff
@@ -0,0 +1,158 @@
+import enum
+import itertools
+
+import networkx
+
+from common import prefix
+from shortest_common_superstring.shortest_common_superstring import naive
+from shortest_common_superstring.teng_yao import cycle_cover
+
+VertexType = enum.IntEnum('VertexType', 'IN OUT HEAD TAIL JOIN')
+
+def _cycle_edges(cycle):
+  return zip(cycle, cycle[1:] + cycle[:1])
+
+def _path_weight(matrix, cycle):
+  return sum(matrix[i][j] for i, j in zip(cycle, cycle[1:]))
+
+def max_weight_perfect_matching(edges):
+  """Find maximum weight perfect matching.
+  Input edges are given as list of triples (vertex, vertex, edge weight).
+  Returns matching as set of pairs (vertex, vertex)."""
+  graph = networkx.Graph()
+  graph.add_weighted_edges_from(edges)
+  return networkx.max_weight_matching(graph, maxcardinality=True)
+
+def _half_edge_gadget(a, b, weight):
+  """Half of gadget corresponding to a 2-cycle on vertices (a, b).
+  Refer to Figures 1 and 2 in the original paper for explanation."""
+  out_a, in_b = (VertexType.OUT, a), (VertexType.IN, b)
+  head_ab, tail_ab = (VertexType.HEAD, a, b), (VertexType.TAIL, a, b)
+  head_ba, join_ab = (VertexType.HEAD, b, a), (VertexType.JOIN, a, b)
+  return [
+    (out_a, tail_ab, weight),
+    (tail_ab, head_ab, 0),
+    (head_ab, in_b, weight),
+    (join_ab, tail_ab, 0),
+    (join_ab, head_ba, 0),
+  ]
+
+def _is_half_edge_matched(matching, a, b):
+  """Check if either of half-edges (a, v(a,b)) or (v(a,b), b) is matched."""
+  out_a, in_b = (VertexType.OUT, a), (VertexType.IN, b)
+  head_ab, tail_ab = (VertexType.HEAD, a, b), (VertexType.TAIL, a, b)
+  return (out_a, tail_ab) in matching or (in_b, head_ab) in matching or \
+         (tail_ab, out_a) in matching or (head_ab, in_b) in matching
+
+def max_cycle_cover_with_half_edges(matrix):
+  """Find maximum weight cycle cover without 2-cycles but with half-edges.
+  If both (u, v) and (v, u) is in returned set of edges,
+  then the edge is undirected (it has two tails or two heads)."""
+  n, auxiliary_graph = len(matrix), []
+  for i, j in itertools.product(range(n), repeat = 2):
+    if i != j:
+      auxiliary_graph += _half_edge_gadget(i, j, matrix[i][j])
+  matching = max_weight_perfect_matching(auxiliary_graph)
+  cover = set()
+  for i, j in itertools.product(range(n), repeat = 2):
+    if i != j and _is_half_edge_matched(matching, i, j):
+      cover.add((i, j))
+  return cover
+
+def _get_cycles(cover):
+  """Iterate over weakly connected components of the cover,
+  which are cycles if we ignore the edge direction."""
+  neighbours, visited = {}, set()
```

Comment:
Consider using for neighbours:
- import collections
- neighbours = collections.defaultdict(set)
This way you do not need L67-68 I guess